llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -4565,6 +4607,11 @@ Instruction *InstCombinerImpl::foldICmpBinOp(ICmpInst &I,
   if (Instruction *NewICmp = foldICmpXNegX(I, Builder))
     return NewICmp;
 
+  if (BO0) {
```

Comment:
Could you please add a test that commutes the operands of icmp? I don't think the complexity-based canonicalization always places `B` in RHS.
```
define i1 @icmp_common_add_sub_commuted(i32 %a, i32 %bb, i32 %c){
   %b = mul %bb, %bb ; thwart complexity-based canonicalization
   %add = add i32 %a, %b
   %sub = sub i32 %add, %c
   %cmp = icmp eq i32 %b, %sub
   ret i1 %cmp
 }
```